### PR TITLE
rewrite barnacle

### PR DESCRIPTION
**Pull request by adalpane662**
## Summary by CodeRabbit

* **New Features**
  * Introduced a unified, builder-based middleware for rate limiting and API key validation, allowing flexible configuration and custom validator support.
  * Added comprehensive unit tests for middleware and in-memory store logic, covering edge cases and concurrency.

* **Refactor**
  * Replaced legacy API key middleware and helper functions with a new generic middleware and builder pattern.
  * Simplified example usage and updated documentation to reflect the new API and builder approach.
  * Updated configuration types and removed deprecated or redundant fields and methods.
  * Removed combined check-and-increment method from Redis store implementation.

* **Bug Fixes**
  * Improved test isolation and reliability by parameterizing routes and enhancing Redis key management in tests.

* **Chores**
  * Updated dependencies and package version.
  * Enhanced `.gitignore` to exclude IDE-specific files.

* **Documentation**
  * Extensively revised README and crate-level docs for clarity and alignment with the new middleware API.

### CODE REVIEW

File: src/middleware.rs [hunk 1]
```diff
@@ -486,24 +240,279 @@ fn get_fallback_key_common(
...
+                    match state {
+                        Some(state) => {
+                            validator.call(api_key.to_string(), api_key_config, Arc::new(parts.clone()), state).await
+                        }
+                        None => {
+                            // Return a more appropriate error for missing validator state
+                            Err(E::from(BarnacleError::custom("Barnacle: API key validator requires state, but none was provided. Use with_state() or use () for stateless validators.", None)))
+                        }
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (adalpane662) [reply src/middleware.rs:402]: how do you suggest to solve this ? middleware must know the type of State to pass into validator. User may provide whatever type of state they want. We cant oblige to provide a default state
